### PR TITLE
[SPARK-35815][SQL] Allow delayThreshold for watermark to be represented as ANSI interval literals

### PR DESCRIPTION
**Pull request by sarutak**
### What changes were proposed in this pull request?

This PR extends the way to represent `delayThreshold` with ANSI interval literals for watermark.

### Why are the changes needed?

A `delayThreshold` is semantically an interval value so it's should be represented as ANSI interval literals as well as the conventional `1 second` form.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

New tests.

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala [hunk 1]
```diff
@@ -739,13 +740,23 @@ class Dataset[T] private[sql](
...
-    val parsedDelay =
-      try {
+    val parsedDelay = try {
+      if (delayThreshold.toLowerCase(Locale.ROOT).startsWith("interval")) {
+        CatalystSqlParser.parseExpression(delayThreshold) match {
+          case Literal(months: Int, _: YearMonthIntervalType) =>
+            new CalendarInterval(months, 0, 0)
+          case Literal(micros: Long, _: DayTimeIntervalType) =>
+            val days = micros / DateTimeConstants.MICROS_PER_DAY
+            val restMicros = micros % DateTimeConstants.MICROS_PER_DAY
```
Comment (MaxGekk) [inline sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala:750]: What's the reason to perform such normalization? Why cannot you just pass `micros` as the third parameter?

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/EventTimeWatermarkSuite.scala [hunk 2]
```diff
@@ -765,6 +765,55 @@ class EventTimeWatermarkSuite extends StreamTest with BeforeAndAfter with Matche
...
+          .withWatermark("eventTime", delayThreshold)
+        val eventTimeAttr = df.queryExecution.analyzed.output.find(a => a.name == "eventTime")
+        assert(eventTimeAttr.isDefined)
+        val metadata = eventTimeAttr.get.metadata
+        assert(metadata.contains(EventTimeWatermark.delayKey))
+        assert(metadata.getLong(EventTimeWatermark.delayKey) === expectedMs)
+      }
+    }
+
+    // Invalid interval patterns
```
Comment (MaxGekk) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/EventTimeWatermarkSuite.scala:800]: Should we allow an interval which begins from gaps like `  interval `?

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/EventTimeWatermarkSuite.scala [hunk 3]
```diff
@@ -765,6 +765,55 @@ class EventTimeWatermarkSuite extends StreamTest with BeforeAndAfter with Matche
     }
   }
 
+  test("SPARK-35815: Support ANSI intervals for delay threshold") {
+    Seq(
+      // Conventional form and some variants
+      (Seq("3 days", "Interval 3 day", "inTerval '3' day"), 259200000),
+      (Seq("5 hours", "INTERVAL 5 hour", "interval '5' hour"), 18000000),
+      (Seq("8 minutes", "interval 8 minute", "interval '8' minute"), 480000),
+      (Seq("10 seconds", "interval 10 second", "interval '10' second"), 10000),
```
Comment (MaxGekk) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/EventTimeWatermarkSuite.scala:774]: Could you replace the magic numbers via constants like `10 * MILLIS_PER_SECOND` Now it is hard to check correctness of the expected values. Especially, the values below.

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala (hunk 1 above)
Comment (sarutak) [reply sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala:750]: Ah, we don't need to normalize. Thanks.

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/EventTimeWatermarkSuite.scala (hunk 2 above)
Comment (sarutak) [reply sql/core/src/test/scala/org/apache/spark/sql/streaming/EventTimeWatermarkSuite.scala:804]: Thanks. Changed to consider such cases.

**Comment by HyukjinKwon** [comment]
cc @HeartSaVioR and @gengliangwang FYI

File: common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/DateTimeConstants.java [hunk 4]
```diff
@@ -21,6 +21,8 @@
 
   public static final int MONTHS_PER_YEAR = 12;
 
+  public static final byte DAYS_PER_MONTH = 31;
```
Comment (MaxGekk) [inline common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/DateTimeConstants.java:24]: Let's **not** add this here. 31 days per month is streaming specific constant. And it is not applicable to other classes in  Spark SQL.

**Review by MaxGekk — APPROVED** [review]

**Comment by MaxGekk** [comment]
+1, LGTM. Merging to master/3.2.
Thank you, @sarutak .

**Comment by HeartSaVioR** [comment]
Late +1. Thanks for dealing with this!

**Pull request closed without merging**